### PR TITLE
Increasing memory and CPU limits for core/critical components

### PR DESCRIPTION
**Pull request by ggaurav10**
Increasing memory and CPU limits for kube-proxy, ingress-controller, calico and etcd based on observations from performance tests.

**What this PR does / why we need it**:
For network heavy workloads, the limits assigned to kube-proxy, ingress controller weren't enough. Thus need to be increased.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
The "requests" values are kept the same, and might be changed as part of another PR.
Idea is to launch pods so that they consume/reserve less resources to begin with. The limits are increased so that those critical components can grow if necessary.

**Release note**:
```noteworthy user
The resource limits for kube-proxy, calico, etcd and nginx-ingress have been increased. Also, the eviction thresholds for available memory on the nodes is now optimized for the used machine type.
```

### CODE REVIEW

**Comment by rfranzke** [comment]
@ggaurav10 3 days ago
>  First, It needs to be under worker because we generate cloud-config secret for each worker separately, and the thresholds might be different for different workers. Under worker i can create further "structures" as you suggested?
> WDYT?

@rfranzke 2 days ago
> yes, sounds good

**Comment by rfranzke** [comment]
@vlerenc commented a day ago
> In case you wonder why I ask @ggaurav10 @amshuman-kr, here my thoughts:
> - We are modifying something dangerous here, so let's be certain that what we do really improves the situation - in all cases, e.g. for all machine types, small and large. We can not afford destabilisations that harm our reputation.
> - Are we certain that `kube-reserved`, https://github.com/gardener/gardener/blob/master/charts/shoot-cloud-config/charts/original/templates/kubelet/_kubelet.flags#L35 and https://github.com/gardener/gardener/blob/master/charts/shoot-cloud-config/charts/original/templates/kubelet/_kubelet-config-file#L66-L68, are right? Maybe we have not only or not at all an eviction, but a misconfiguration problem here? `docker`, `kubelet`, other system processes have enough resources?
> - It would be quick to test I guess:
>   - Have a small machine
>   - Have a large machine
>   - Take a look at the memory in use without pods and with 110 pods loaded
> - Larger machines that run more pods may mean, that things average out (only 1-2 of 30-40 spike up)
> - On the other hand, it could also mean, that more pods exceed their allotted memory
> 
> P.S.: We don't want the OOM killer to go around shooting down processes, just because too many pods spike and the kubelet is too lazy to shoot them down and then processes we don't want shot get killed. If we only increase the memory in all cases, then OK, but we should take the day for something like above to get a good grip on the memory footprint.

**Comment by ggaurav10** [comment]
@vlerenc 
let's have a look at a typical example:

Node: Memory: 60GB
Hard eviction: 1GB (previous value 100MB)
kube-reserved: 1.5GB

This leaves allocatable: 57.5GB (instead of previous value of 58.4GB) for the machine.
For small machines of 8GB memory, this value would be 6.1GB under new values, instead of previous 6.4GB (hard eviction at 400MB).
This would provide some room for system daemons to grow, and not be under pressure. This does reduce allocatable memory to user pods by about 400MB (for 8GB machine) and 1GB (for 60GB machine). IMHO, this is a small price to pay for ensuring that system critical components can still run when under pressure.
Alternatively, we can look into the kubelet flag `system-reserved` too.

We will gather some more data from perf tests as you suggested, and keep iterating the configurations if required.

**Comment by ggaurav10** [comment]
Something else we need to look at when collecting data from perf tests on small and large machines, and then configuring them: limiting maximum number of pods on a node.
This is currently configured to 110 for all sizes of machines. 
Smaller nodes with same number of pods as larger machines would start seeing some resource pressure (even evictions) if the resource consumption of workloads increases once already deployed.

**Comment by rfranzke** [comment]
We agreed on merging these changes for now as they improve the situation, but the exact values still need a little tweaking. This will happen in an iterative way.

As discussed, let's please remove https://github.com/gardener/gardener/blob/f6f8e2382143d9179d8ec13657515e46ffc97669/pkg/operation/common/types.go#L217 and all occurrences, e.g.  https://github.com/gardener/gardener/blob/f6f8e2382143d9179d8ec13657515e46ffc97669/pkg/operation/cloudbotanist/awsbotanist/cloud_config.go#L24 before merging this.

**Comment by ggaurav10** [comment]
As suggested by @rfranzke, i have removed the unused field. However, i have left the function `GetWorkerNames()` untouched.

**Review by zanetworker — DISMISSED** [review]

File: pkg/operation/hybridbotanist/cloud_config.go [hunk 1]
```diff
@@ -55,12 +56,29 @@ func (b *HybridBotanist) generateCloudConfigChart() (*chartrenderer.RenderedChar
...
+		for _, machtype := range machineTypes {
+			if machtype.Name == worker.MachineType {
+				if machtype.Memory.Cmp(memoryThreshold) > 0 {
+					newWorker["evictionHard"] = "1Gi"
+					newWorker["evictionSoft"] = "1.5Gi"
+				} else {
+					newWorker["evictionHard"] = "5%"
+					newWorker["evictionSoft"] = "10%"
+				}
+				workers = append(workers, newWorker)
```
Comment (rfranzke) [inline pkg/operation/hybridbotanist/cloud_config.go]: Shouldn't this be moved to line 82? I see that the code would work as is but this `append` is quite confusing here. Also, the `break` is would no longer be needed after moving the statement (if there are multiple machine types with the same name then simply the last one would "win").

File: pkg/operation/hybridbotanist/cloud_config.go [hunk 2]
```diff
@@ -55,12 +56,29 @@ func (b *HybridBotanist) generateCloudConfigChart() (*chartrenderer.RenderedChar
...
+
+	for _, worker := range b.Shoot.GetWorkers() {
+		newWorker := map[string]interface{}{
+			"name":       worker.Name,
+			"secretName": b.Shoot.ComputeCloudConfigSecretName(worker.Name),
+		}
+		for _, machtype := range machineTypes {
+			if machtype.Name == worker.MachineType {
+				if machtype.Memory.Cmp(memoryThreshold) > 0 {
+					newWorker["evictionHard"] = "1Gi"
```
Comment (rfranzke) [inline pkg/operation/hybridbotanist/cloud_config.go]: Could we rename these fields to `eviction{Soft,Hard}MemoryAvailable` to make clear what these values are used for?

**Review by rfranzke — CHANGES_REQUESTED** [review]
Could you please add the new possible values (`eviction{Hard,Soft}MemoryAvailable`) to this section: https://github.com/gardener/gardener/blob/master/charts/shoot-cloud-config/charts/original/values.yaml#L22-L26 (to make clear what is configurable)?

File: pkg/operation/hybridbotanist/cloud_config.go (hunk 1 above)
Comment (ggaurav10) [reply pkg/operation/hybridbotanist/cloud_config.go]: Is it possible to have multiple machine types (different combination of cpu,memory, gpu, etc) with same name?
Even if it is, then the whole logic of comparing memory size values from a given type wouldn't work, right? And we could end up in situations where we reserve 6GB for 60GB machines (instead of 1.5GB). We would need to change the whole logic.
Also, `break` statement stops unnecessary comparisons once a match is found.
WDYT?

File: pkg/operation/hybridbotanist/cloud_config.go (hunk 2 above)
Comment (ggaurav10) [reply pkg/operation/hybridbotanist/cloud_config.go]: Done.

File: pkg/operation/hybridbotanist/cloud_config.go (hunk 1 above)
Comment (rfranzke) [reply pkg/operation/hybridbotanist/cloud_config.go]: > Is it possible to have multiple machine types (different combination of cpu,memory, gpu, etc) with same name?

Yes, there is no validation (yet?) that would prevent this. Agreed that allowing this does not make sense at all (also the quota validation logic would not work as expected). I will add this kind of validation separately (EDIT: FYI, uniqueness validation implemented with 281ed3cce7efebfbca8bf708f60abd5ec3a40548).

Regarding the `append`: What I dislike is that it is only executed if the machine type is found, otherwise the worker configuration won't be added to the cloud-config. Let's assume the `machineTypes := b.Shoot.GetMachineTypesFromCloudProfile()` statement is buggy and returns an empty list - then, in the next reconciliation no `append` statements would be executed any more. We would end up in a situation without ANY worker configuration. If we would move the `append` below the `for`-loop we would "just" misconfigure the eviction parameters. I'm fine with the `break` to save unneeded iterations, but logically IMHO the `append` needs to be moved down.

Also, you have not yet added the new parameters to the `values.yaml` file of the Helm chart:
> Could you please add the new possible values (`eviction{Hard,Soft}MemoryAvailable`) to this section: https://github.com/gardener/gardener/blob/master/charts/shoot-cloud-config/charts/original/values.yaml#L22-L26 (to make clear what is configurable)?

I guess the PR is ready to be merged after these changes, so please squash the commits appropriately.

Thank you! 👍

Comment (ggaurav10) [reply pkg/operation/hybridbotanist/cloud_config.go]: I have changed the logic such that there is always some default values provided for the eviction parameters. These values get replaced in the next `for` loop. This way, even if the statement `machineTypes := b.Shoot.GetMachineTypesFromCloudProfile()` is buggy, we will assign default values.
Does that make sense?

**Review by rfranzke — APPROVED** [review]
/lgtm
